### PR TITLE
*:support aggregate function push cross outer join.

### PR DESCRIPTION
**Pull request by hanfei1991**
Like "select sum(b.c) from a left join b on a.id = b.id", It can be rewrited as "select sum(agg_0) from a left join (select sum(c), id from b group by id) b on a.id = b.id". But instead of null padding, we shoule set sum(c)'s default value as 0.
@shenli @coocood @zimulala @XuHuaiyu PTAL

### CODE REVIEW

**Comment by zimulala** [comment]
go test failed, aggregate_test.go:128
... obtained string = "[[1 1] [10 <nil>] [4 3]]"
... expected string = "[[10 <nil>] [1 1] [4 3]]"

File: executor/executor.go [hunk 1]
```diff
@@ -495,20 +495,21 @@ func init() {
...
+	prepared      bool
+	ctx           context.Context
+	smallFilter   expression.Expression
+	bigFilter     expression.Expression
+	otherFilter   expression.Expression
+	schema        expression.Schema
+	outer         bool
+	leftSmall     bool
+	cursor        int
+	defaultValues []types.Datum
```
Comment (coocood) [inline executor/executor.go]: It's unclear what does `defaultValues` mean in a `HashJoinExec`.
Change it to a more specific name would be better.

File: expression/aggregation.go [hunk 2]
```diff
@@ -74,6 +74,10 @@ type AggregationFunction interface {
 
 	// GetType gets field type of aggregate function.
 	GetType() *types.FieldType
+
+	// GetDefaultValue gets the default value when the aggregate function's input is null.
+	// The input stands for the schema of Aggregation's child.
+	GetDefaultValue(schema Schema) (types.Datum, error)
```
Comment (coocood) [inline expression/aggregation.go]: `GetDefaultValue` is a name too general for what it actually does.

File: expression/expression.go [hunk 3]
```diff
@@ -488,3 +488,28 @@ func SplitCNFItems(onExpr Expression) []Expression {
 func SplitDNFItems(onExpr Expression) []Expression {
 	return splitNormalFormItems(onExpr, ast.OrOr)
 }
+
+// EvaluateExprWithNull sets inner table columns in a expression as null and calculate the final result of the scalar function.
+func EvaluateExprWithNull(schema Schema, expr Expression) (Expression, error) {
```
Comment (coocood) [inline expression/expression.go]: Can we use the standard `Eval` method with a null filled row?

Comment (hanfei1991) [inline expression/expression.go]: It's better not to do that.

File: expression/expression.go [hunk 4]
```diff
@@ -488,3 +488,28 @@ func SplitCNFItems(onExpr Expression) []Expression {
 func SplitDNFItems(onExpr Expression) []Expression {
 	return splitNormalFormItems(onExpr, ast.OrOr)
 }
+
+// EvaluateExprWithNull sets inner table columns in a expression as null and calculate the final result of the scalar function.
```
Comment (coocood) [inline expression/expression.go]: `sets columns in the schema as null`
This function doesn't know if `schema` is inner table.

**Comment by hanfei1991** [comment]
@coocood @zimulala @XuHuaiyu PTAL

File: expression/expression.go [hunk 5]
```diff
@@ -488,3 +488,29 @@ func SplitCNFItems(onExpr Expression) []Expression {
...
+
+// EvaluateExprWithNull sets columns in schema as null and calculate the final result of the scalar function.
+// If the Expression is a non-constant value, it means the result is unknown.
+func EvaluateExprWithNull(schema Schema, expr Expression) (Expression, error) {
+	switch x := expr.(type) {
+	case *ScalarFunction:
+		var err error
+		args := make([]Expression, len(x.Args))
+		for i, arg := range x.Args {
+			args[i], err = EvaluateExprWithNull(schema, arg)
```
Comment (zimulala) [inline expression/expression.go]: The `err` will be overwritten.

File: expression/aggregation.go [hunk 6]
```diff
@@ -321,6 +330,20 @@ func (sf *sumFunction) GetStreamResult() (d types.Datum) {
...
+// CalculateDefaultValue implements AggregationFunction interface.
+func (sf *sumFunction) CalculateDefaultValue(schema Schema) (d types.Datum, err error) {
+	arg := sf.Args[0]
+	result, err := EvaluateExprWithNull(schema, arg)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	if con, ok := result.(*Constant); ok {
+		d, err = types.CalculateSum(d, con.Value)
+		return
```
Comment (zimulala) [inline expression/aggregation.go]: Add trace.

File: plan/aggregation_push_down.go [hunk 7]
```diff
@@ -224,9 +225,28 @@ func (a *aggPushDownSolver) tryToPushDownAgg(aggFuncs []expression.AggregationFu
 	}
 	agg.AggFuncs = newAggFuncs
 	agg.SetSchema(schema)
+	if (childIdx == 0 && join.JoinType == RightOuterJoin) || (childIdx == 1 && join.JoinType == LeftOuterJoin) {
+		var err error
+		join.DefaultValues, err = a.getDefaultValues(agg)
+		if err != nil {
```
Comment (zimulala) [inline plan/aggregation_push_down.go]: If the `err` doesn't return, it need to add some log information.

**Comment by hanfei1991** [comment]
@coocood @zimulala PTAL

**Comment by coocood** [comment]
LGTM

File: plan/logical_plans.go [hunk 8]
```diff
@@ -51,6 +52,10 @@ type Join struct {
 	LeftConditions  []expression.Expression
 	RightConditions []expression.Expression
 	OtherConditions []expression.Expression
+
+	// DefaultValues only used for outer join, which stands for the default values when the outer table cannot find join partner
```
Comment (XuHuaiyu) [inline plan/logical_plans.go]: only used --> is only used

File: expression/aggregation.go [hunk 9]
```diff
@@ -321,6 +332,24 @@ func (sf *sumFunction) GetStreamResult() (d types.Datum) {
 	return
 }
 
+// CalculateDefaultValue implements AggregationFunction interface.
+func (sf *sumFunction) CalculateDefaultValue(schema Schema) (d types.Datum, valid bool) {
+	arg := sf.Args[0]
+	result, err := EvaluateExprWithNull(schema, arg)
+	if err != nil {
+		log.Warn(err.Error())
```
Comment (zimulala) [inline expression/aggregation.go]: Add more information in log. The same as line-346.

File: expression/expression.go [hunk 10]
```diff
@@ -488,3 +488,29 @@ func SplitCNFItems(onExpr Expression) []Expression {
...
+				return nil, errors.Trace(err)
+			}
+		}
+		return NewFunction(x.FuncName.L, types.NewFieldType(mysql.TypeTiny), args...)
+	case *Column:
+		if schema.GetIndex(x) == -1 {
+			return x, nil
+		}
+		constant := &Constant{Value: types.Datum{}}
+		constant.Value.SetNull()
```
Comment (zimulala) [inline expression/expression.go]: Remove this unnecessary line.

**Pull request merged**